### PR TITLE
Errors occurring up to the end of setup are now collected and raised as a single exception.

### LINKED ISSUE #2701: Failure of shape by connection causes errors that do not allow the N2 to be built.
### Description

When generating an N2, OpenMDAO attempts to allow errors to pass so that the N2 can be generated.  A user has reported that errors caused by `shape_by_conn` interrupt this process and prevent the N2 generation.

We need to reproduce this behavior and make sure that we allow the N2 to be generated when these errors are present.

### Example

```python
TBD
```

### OpenMDAO Version

3.21.1-dev

### Relevant environment information

_No response_

### LINKED ISSUE #2702: Save all connection errors and report them at once.
### Desired capability or behavior.

Currently OpenMDAO generates connection errors and fails at the first possible chance.  Users rightly are frustrated by working through them one by one by one.  OpenMDAO's behavior should be such that these errors are cached and dumped out all at the same time.

### Associated POEM

None

### PR DESCRIPTION
**Pull request by naylor-b**
### Summary

Also modified the reports system so that it can handle more complex hook configurations.  The N2 now uses 2 hooks, with a hook just before processing the collected errors that will run the N2 if any errors have occurred, and a later post final_setup hook that will run the N2 in the error free case.

Also added a check for the environment variable OPENMDAO_FAIL_FAST.  If that env var is 'truthy', then the first error encountered prior to the end of setup will be raised as an exception immediately. This exists primarily to make tracking down the original error easier when debugging.

### Related Issues

- Resolves #2702
- Resolves #2701

### Backwards incompatibilities

Existing unit tests could break if they are dependent upon the previous error behavior.  Users will now see different error messages, sometimes more verbose than they did previously, but models that weren't already broken should not break due to this PR.

### New Dependencies

None

### CODE REVIEW

File: openmdao/utils/reports_system.py [hunk 1]
```diff
@@ -80,8 +115,69 @@ def register_report(name, func, desc, class_name, method, pre_or_post, filename=
...
+    fname : str
+        The name of the function where the pre and/or post hook will be applied.
+    class_name : str
+        The name of the class owning the method where the hook will be applied.
+    inst_id : str or None
+        The name of the instance owning the method where the hook will be applied.
+    pre : function (None)
+        If not None, this hook will run before the named function runs.
+    post : function (None)
+        If not None, this hook will run after the named function runs.
```
Comment (swryan) [inline openmdao/utils/reports_system.py:161]: pedantic nit...  these arguments are a function _reference_ not a function name (so "after the referenced function"...)

Also,  should this function be added to the doc page for the Reports System?

File: openmdao/core/problem.py [hunk 2]
```diff
@@ -774,6 +779,32 @@ def _set_initial_conditions(self):
...
+        if self._metadata['saved_errors'] is None:
+            return
+
+        errors = self._metadata['saved_errors']
+
+        # set the errors to None so that all future calls will immediately raise an exception.
+        self._metadata['saved_errors'] = None
+
+        if errors:
+            final_msg = [f"\nConnection errors for problem '{self._name}':"]
```
[the remark this answers is not in the document]
Comment (naylor-b) [reply openmdao/core/problem.py]: Changed these to "Collected errors..."

**Review by swryan — APPROVED** [review]

File: openmdao/core/system.py [hunk 3]
```diff
@@ -1609,12 +1646,6 @@ def _setup_global_connections(self, conns=None):
         """
         Compute dict of all connections between this system's inputs and outputs.
 
-        The connections come from 4 sources:
-        1. Implicit connections owned by the current system
```
[the remark this answers is not in the document]
Comment (naylor-b) [reply openmdao/core/system.py:1613]: It's technically accurate, but there could be some confusion due to the use of 'implicit' here.  The top line of the docstring says this computes the dict for *all* connections, and I just didn't feel that this added detail was really helpful.  Also, this particular version of the method is a no-op.

File: openmdao/utils/reports_system.py [hunk 4]
```diff
@@ -80,8 +166,66 @@ def register_report(name, func, desc, class_name, method, pre_or_post, filename=
...
-                                      filename)
+    _reports_registry[name] = report = Report(name, desc)
+
+    pre = func if pre_or_post == 'pre' else None
+    post = func if pre_or_post == 'post' else None
+    report.register_hook_args(fname=method, class_name=class_name, inst_id=inst_id, pre=pre,
+                              post=post, ncalls=1, **kwargs)
+
+
+def unregister_report(name):
```
[the remark this answers is not in the document]
Comment (naylor-b) [reply openmdao/utils/reports_system.py:177]: fixed

**Review by Kenneth-T-Moore — APPROVED** [review]

**Pull request merged**